lib/grape/middleware/versioner/header.rb

```diff
@@ -23,53 +23,83 @@ module Versioner
-            if header.qvalues.empty?
-              fail Grape::Exceptions::InvalidAcceptHeader.new('Accept header must not contain ranges ("*").',
-                                                              error_headers)
-            end
-          end
-
-          media_type = header.best_of available_media_types
+          strict_header_checks if strict?
 
           if media_type
-            type, subtype = Rack::Accept::Header.parse_media_type media_type
-            env['api.type']    = type
-            env['api.subtype'] = subtype
-
-            if /\Avnd\.([a-z0-9*.]+)(?:-([a-z0-9*\-.]+))?(?:\+([a-z0-9*\-.+]+))?\z/ =~ subtype
-              env['api.vendor']  = Regexp.last_match[1]
-              env['api.version'] = Regexp.last_match[2]
-              env['api.format']  = Regexp.last_match[3]  # weird that Grape::Middleware::Formatter also does this
-            end
+            media_type_header_handler
           # If none of the available content types are acceptable:
           elsif strict?
-            fail Grape::Exceptions::InvalidAcceptHeader.new('406 Not Acceptable', error_headers)
-          # If all acceptable content types specify a vendor or version that doesn't exist:
-          elsif header.values.all? { |header_value| has_vendor?(header_value) || version?(header_value) }
-            fail Grape::Exceptions::InvalidAcceptHeader.new('API vendor or version not found.', error_headers)
+            fail_with_invalid_accept_header('406 Not Acceptable')
+          elsif headers_contain_wrong_vendor_or_version?
+            fail_with_invalid_accept_header('API vendor or version not found.')
           end
         end
 
         private
 
+        def strict_header_checks
+          # If no Accept header:
+          if header.qvalues.empty?
+            fail_with_invalid_accept_header('Accept header must be set.')
+          end
+          # Remove any acceptable content types with ranges.
+          header.qvalues.reject! do |media_type, _|
+            Rack::Accept::Header.parse_media_type(media_type).find do |s|
+              s == '*'
+            end
+          end
+          # If all Accept headers included a range:
+          if header.qvalues.empty?
+            fail_with_invalid_accept_header(
+              'Accept header must not contain ranges ("*").'
+            )
+          end
+        end
+
+        def header
+          @header ||= rack_accept_header
+        end
+
+        def media_type
+          @media_type ||= header.best_of available_media_types
+        end
+
+        def media_type_header_handler
+          type, subtype = Rack::Accept::Header.parse_media_type media_type
+          env['api.type']    = type
+          env['api.subtype'] = subtype
+
+          if vendor_version_header_regex =~ subtype
+            env['api.vendor']  = Regexp.last_match[1]
+            env['api.version'] = Regexp.last_match[2]
+            # weird that Grape::Middleware::Formatter also does this
```

Comment:
Open an issue with a link to this code to figure out why Grape::Middleware::Formatter does this instead.